### DISCUSSION TITLE
More Pixel Art Upscaling Methods

### DISCUSSION THREAD

**Discussion by drummyfish** [ImageMagick/ImageMagick#6698] [Contribute]
Hello,

I've been unsuccessfully looking for a tool that implements pixel art scaling methods and so I thought I could write one and integrate it into ImageMagick. AFAIK it currently supports scale2x via [-magnify](https://www.imagemagick.org/script/command-line-options.php#magnify) option, is that right? I'd like to ask:

- Should I? I.e. isn't anyone already working on this, is there an existing tool that I missed etc.?
- What should the CLI options look like? Maybe `-magnify [method]` where `method` could be `eagle2x`, `hq3x` etc.?

The closest thing to what I wanted that I found was [ImageResizer](https://forums.rpgmakerweb.com/index.php?threads/image-resizer-specially-for-upscaling-pixel-art.45856/) but it's written in .NET and doesn't run under mono/wine for me. I'd also prefer a CLI tool. The source of ImageResizer could still be useful to rewrite the algorithms from.

Sorry if I'm posting in a wrong place, I'm new.

EDIT:

Additional questions/thoughts:
- As I'm looking at the code, the scale2x algorithm is very hardcoded and copy/pasted, only allows 2x upscaling and intensity-based pixel comparison. This will have to be rewritten to probably have a "master" function that will iterate on pixels and then have a set of "slave" functions for each upscaling algorithm to compute the pixel color based on the neighbouring pixels. If anyone has an idea how to do this nicely, let me know.

Also for reference, [here](https://github.com/drummyfish/pixelart-upscale/blob/master/upscale.py) I have some of the algorithms I have written in Python some time ago.

**Comment by urban-warrior** [comment]
ImageMagick currently only supports scale2x.  If you want to implement other methods, you will need a new command line option, perhaps -pixel-art method.  If you need pixel art options, utilize defines.  Something like -define pixel-art:threshold=50%.

**Comment by drummyfish** [comment]
Then it would perhaps be best to keep the `-magnify` option and use the `-define` to specify the algorithm? Would make sense.

**Comment by urban-warrior** [comment]
Agreed, something like -define magnify:method=scale2X  -magnify.
